### ISSUE TITLE
Operator overloading rules changes on 3.0.0

### ISSUE THREAD

**Issue by cquiroz** [lampepfl/dotty#12708]
Porting [spire](https://github.com/typelevel/spire) shows a change (Couldn't find any mention to this in the docs) from Scala 2.13 and 3.0 on operator overloading. This is repeated across the codebase which contains overloads to common operators like `+` and `-`. The code below is a minimization that compiles in 2.13 but fails in 3.0

A repo with the issue can be found at

https://github.com/cquiroz/implicit-operator

## Compiler version
Scala 2.13 and 3.0.0.

## Minimized code

```Scala

import algebra.ring.AdditiveSemigroup
import algebra.ring.CommutativeSemiring
import algebra.Semigroup
import algebra.Monoid

final class AdditiveSemigroupOps[A](lhs: A)(implicit as: AdditiveSemigroup[A]) {
  def +(rhs: A): A = as.plus(rhs, lhs)
  def ^(rhs: A): A = as.plus(rhs, lhs)
}

trait AdditiveSemigroupSyntax {
  implicit def additiveSemigroupOps[A: AdditiveSemigroup](
      a: A
  ): AdditiveSemigroupOps[A] = new AdditiveSemigroupOps(a)
}

object syntax {
  object additiveSemigroup extends AdditiveSemigroupSyntax
}

object App {
  def main(args: Array[String]): Unit = {
    import syntax.additiveSemigroup._
    implicit def IntAlgebra[A]: AdditiveSemigroup[Map[Int, A]] = ???

    def res[A]: Map[Int, A] = {
      val a: Map[Int, A] = Map.empty
      val b: Map[Int, A] = Map.empty
      // Calls the operator on AdditiveSemigroupOps
      a ^ b
      // Calls the operator + on AdditiveSemigroupOps only in Scala 2
      // In Scala 3 tries to call `+` on Map
      a + b
    }
  }

}

```

## Output

```scala
// Only in Scala 3.0
[error] -- [E007] Type Mismatch Error: /Users/cquiroz/Projects/implicit-operator/src/main/scala/Main.scala:33:10
[error] 33 |      a + b
[error]    |          ^
[error]    |          Found:    (b : Map[Int, A])
[error]    |          Required: (Int, A)

```

## Expectation

To cross compile or have some way to override de `+` operator like with the `^` operator.

**Comment by som-snytt** [comment]
Reducing the mechanism:
```
➜  snips cat icq.scala

type MIA = Map[Int, Any]

extension (m: MIA) def +(other: MIA): MIA = ???

def f(mia: MIA) = mia + mia
➜  snips scalac icq.scala
-- [E007] Type Mismatch Error: icq.scala:6:24 ------------------------------------------------------------------------
6 |def f(mia: MIA) = mia + mia
  |                        ^^^
  |                        Found:    (mia : MIA)
  |                        Required: (Int, Any)

longer explanation available when compiling with `-explain`
1 error found
```
Worth noting that `MapOps` has params `C`, `CC`.
```
➜  snips scalac -Ydebug -Ylog:all icq.scala
...
[log typer] cycle detected for C, true, true
[log typer] cycle detected for CC, true, true
[log typer] cycle detected for C, true, true
[log typer] cycle detected for CC, true, true
[log typer] cycle detected for C, true, true
[log typer] cycle detected for C, true, true
-- [E007] Type Mismatch Error: icq.scala:6:24 ------------------------------------------------------------------------
6 |def f(mia: MIA) = mia + mia
  |                        ^^^
  |                        Found:    (mia : MIA)
  |                        Required: (Int, Any)
```
Not sure what other output knobs to try.

**Comment by bishabosha** [comment]
version without spire dependency:
```scala
import language.implicitConversions

trait AdditiveSemigroup[A]

final class AdditiveSemigroupOps[A](lhs: A)(implicit as: AdditiveSemigroup[A]) {
  def +(rhs: A): A = ???
  def ^(rhs: A): A = ???
}

trait AdditiveSemigroupSyntax {
  implicit def additiveSemigroupOps[A: AdditiveSemigroup](a: A): AdditiveSemigroupOps[A] =
     new AdditiveSemigroupOps(a)
}

object syntax {
  object additiveSemigroup extends AdditiveSemigroupSyntax
}

object App {

  def main(args: Array[String]): Unit = {
    import syntax.additiveSemigroup._

    implicit def IntAlgebra[A]: AdditiveSemigroup[Map[Int, A]] = ???

    def res[A]: Map[Int, A] = {
      val a: Map[Int, A] = Map.empty
      val b: Map[Int, A] = Map.empty
      // Calls the operator on AdditiveSemigroupOps
      a ^ b
      // Calls the operator + on AdditiveSemigroupOps only in Scala 2
      // In Scala 3 tries to call `+` on Map
      a + b
    }
  }

}
```

**Comment by som-snytt** [comment]
https://dotty.epfl.ch/docs/reference/changed-features/implicit-conversions.html says it doesn't try conversions for the 3rd case in Scala 2 views https://scala-lang.org/files/archive/spec/2.13/07-implicits.html#views where there is a member but the args are wrong.

**Comment by som-snytt** [comment]
Adjacently, https://github.com/lampepfl/dotty/issues/3769 notes that `e.m[A]` is not subject to conversion.

**Comment by cquiroz** [comment]
Will the PR allow the conversion as in scala 2?

**Comment by bishabosha** [comment]
> Will the PR allow the conversion as in scala 2?

with the given change `a + b` becomes `additiveSemigroupOps(a) + b`

**Comment by som-snytt** [comment]
Even my simple extension example works!

Showing that scala 2 and 3 `-Xprint` the same, except for a space:
```
syntax.additiveSemigroup.additiveSemigroupOps[Map[Int,A]](a)(IntAlgebra[A]).+(b)
syntax.additiveSemigroup.additiveSemigroupOps[Map[Int, A]](a)(IntAlgebra[A]).+(b)
```

**Comment by cquiroz** [comment]
I tested this on my test project but I still see the same compilation error on 3.0.1-RC1/RC2
Looking at the release notes it seems the fix is not included, is it?

**Comment by dwijnand** [comment]
No, the fix is in master so will be in 3.1.0 (and its pre-releases) - unless it were backported to the https://github.com/lampepfl/dotty/tree/release-3.0.1 branch.
